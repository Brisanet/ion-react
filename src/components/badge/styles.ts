import styled, { DefaultTheme, RuleSet, css } from 'styled-components';
import { BadgeType } from './badge';

type BadgeStylesProps = {
  $type: BadgeType;
};

type ColorByVariant = {
  background: string;
  color?: string;
};

export const typeColors: (
  theme: DefaultTheme,
  type: BadgeType
) => ColorByVariant = ({ colors }, type) =>
  ({
    primary: {
      background: colors.main.primary,
    },
    secondary: {
      background: colors.primary[1],
      color: colors.main.primary,
    },
    neutral: {
      background: colors.neutral[4],
      color: colors.neutral[8],
    },
    danger: {
      background: colors.main.negative,
    },
  }[type]);

const type: (theme: DefaultTheme, type: BadgeType) => RuleSet<object> = (
  theme,
  type
) => {
  const { background, color } = typeColors(theme, type);

  return css`
    background-color: ${background};
    color: ${color || theme.colors.neutral[1]};
  `;
};

export const Badge = styled.span<BadgeStylesProps>`
  ${({ theme, $type }) => css`
    ${theme.font.size[12]};
    font-weight: 600;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    padding: 0 0.4rem;
    border-radius: 50rem;
    user-select: none;
    ${type(theme, $type)};
  `}
`;
