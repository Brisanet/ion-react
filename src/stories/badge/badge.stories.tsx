import { ComponentMeta, ComponentStory } from '@storybook/react';

import { BadgeProps, IonBadge } from '../../components/badge/badge';

export default {
  title: 'Ion/Data display/Badges',
  component: IonBadge,
} as ComponentMeta<typeof IonBadge>;

const Template: ComponentStory<typeof IonBadge> = (args: BadgeProps) => (
  <IonBadge {...args} />
);

export const StringLabel = Template.bind({});
StringLabel.args = {
  label: 'Label',
};

export const NumberLabel = Template.bind({});
NumberLabel.args = {
  label: 10,
};

export const BiggerValue = Template.bind({});
BiggerValue.args = {
  label: 100,
};

export const Primary = Template.bind({});
Primary.args = {
  label: 'Badge',
  type: 'primary',
};

export const Secondary = Template.bind({});
Secondary.args = {
  label: 'Badge',
  type: 'secondary',
};

export const Neutral = Template.bind({});
Neutral.args = {
  label: 'Badge',
  type: 'neutral',
};

export const Danger = Template.bind({});
Danger.args = {
  label: 'Badge',
  type: 'danger',
};
