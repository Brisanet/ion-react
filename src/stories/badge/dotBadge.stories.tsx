import { ComponentMeta, ComponentStory } from '@storybook/react';

import { BadgeProps, IonBadge } from '../../components/badge/badge';

export default {
  title: 'Ion/Data display/Badges/Dot',
  component: IonBadge,
  parameters: {
    controls: { exclude: ['type', 'dot'] },
  },
} as ComponentMeta<typeof IonBadge>;

const Template: ComponentStory<typeof IonBadge> = (args: BadgeProps) => (
  <IonBadge dot={true} {...args} />
);

export const Dot = Template.bind({});
Dot.args = {
  label: 'Badge',
};

export const DotWithIcon = Template.bind({});
DotWithIcon.args = {
  label: 'Badge',
  status: 'negative',
  icon: 'close',
  size: 'md',
};

export const DotWithCustomColor = Template.bind({});
DotWithCustomColor.args = {
  label: 'Badge',
  customColor: '#4d0cb4',
};
