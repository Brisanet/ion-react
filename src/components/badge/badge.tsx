import React, { useState, useEffect } from 'react';
import { BadgeStyles } from './styles';

type StitchesBadgeProps = React.ComponentProps<typeof BadgeStyles>;

export type LabelType = string | number;

export type BadgeType = 'primary' | 'secondary' | 'neutral' | 'negative';

export type BadgeProps = {
  label: LabelType;
  type?: BadgeType;
} & StitchesBadgeProps;

export const IonBadge = ({ label, type = 'primary' }: BadgeProps) => {
  const [badgeValue, setBadgeValue] = useState<LabelType>(0);

  const formatLabel = (badgeLabel: LabelType): LabelType => {
    const parsedToNumber = Number(badgeLabel);
    const maxValue = 99;

    if (isNaN(parsedToNumber)) {
      return badgeLabel;
    }

    return parsedToNumber > maxValue
      ? `${maxValue}+`
      : parsedToNumber.toString();
  };

  useEffect(() => {
    setBadgeValue(formatLabel(label));
  }, [label]);

  return (
    <BadgeStyles type={type} data-testid="ion-badge">
      {badgeValue}
    </BadgeStyles>
  );
};
