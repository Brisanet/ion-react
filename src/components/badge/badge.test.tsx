import React from 'react';
import { render, screen } from '@testing-library/react';

import { IonBadge, BadgeType, LabelType } from './badge';
import { BadgeProps } from './badge';

const defaultBadge: BadgeProps = {
  label: 'Badge',
  type: 'primary',
};

const sut = (props: BadgeProps = defaultBadge) => {
  return render(<IonBadge {...props} />);
};

const getBadge = () => {
  return screen.getByTestId('ion-badge');
};

describe('BadgeComponent', () => {
  describe('Default', () => {
    beforeEach(() => {
      sut();
    });

    it('should render badge component', async () => {
      expect(screen.getByText(defaultBadge.label)).toBeInTheDocument();
    });

    it('should render primary badge by default', async () => {
      expect(getBadge().className).toContain('type-primary');
    });
  });

  describe('Custom Props', () => {
    it.each(['10', 99, '6', 0])(
      'should render badge with %i label (string or number)',
      async (label: LabelType) => {
        sut({ label });
        expect(screen.getByText(label)).toBeInTheDocument();
      }
    );

    it.each([100, 101, 1000])(
      'should render 99+ when value is %i (bigger than 100)',
      async (label: LabelType) => {
        sut({ label });
        expect(getBadge().textContent).toBe('99+');
      }
    );

    it('should not render 99+ if string contains characters other than numbers', async () => {
      const label = '100A';
      sut({ ...defaultBadge, label: label });
      expect(getBadge().textContent).toBe(label);
    });

    const badgeTypes: Array<BadgeType> = [
      'primary',
      'secondary',
      'neutral',
      'negative',
    ];
    it.each(badgeTypes)(
      'should render badge %s type',
      async (type: BadgeType) => {
        sut({ ...defaultBadge, type: type });
        expect(getBadge().className).toContain(`type-${type}`);
      }
    );
  });
});
