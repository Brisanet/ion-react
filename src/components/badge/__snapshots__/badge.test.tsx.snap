// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BadgeComponent Dot Default should render dot badge by default 1`] = `
.c0 {
  width: 8px;
  height: 8px;
  border-radius: 50%;
  background-color: #0858ce;
}

<div
  class="c0"
  data-testid="ion-dot-badge"
/>
`;
